platform/macos/app/MapDocument.m

```diff
@@ -153,6 +157,67 @@ - (NSURL *)shareURL {
              camera.heading, camera.pitch]];
 }
 
+#pragma mark File methods
+
+- (IBAction)takeSnapshot:(id)sender {
+    MGLMapCamera *camera = self.mapView.camera;
+    
+    MGLMapSnapshotOptions* options = [[MGLMapSnapshotOptions alloc] initWithStyleURL:self.mapView.styleURL camera:camera size:self.mapView.bounds.size];
+    options.zoom = self.mapView.zoomLevel;
+    
+    // Create and start the snapshotter
+    snapshotter = [[MGLMapSnapshotter alloc] initWithOptions:options];
+    [snapshotter startWithCompletionHandler: ^(NSImage *image, NSError *error) {
+        if (error) {
+            NSLog(@"Could not load snapshot: %@", [error localizedDescription]);
+        } else {
+            NSWindow* window = [[[self windowControllers] objectAtIndex:0] window];
+    
+            NSString* newName = [[@"snapshot" stringByDeletingPathExtension] stringByAppendingPathExtension:@"png"];
+            
+            // Set the default name for the file and show the panel.
+            NSSavePanel* panel = [NSSavePanel savePanel];
+            [panel setNameFieldStringValue:newName];
+            [panel beginSheetModalForWindow:window completionHandler:^(NSInteger result){
+                if (result == NSFileHandlingPanelOKButton) {
+                    // Write the contents in the new format.
+                    NSURL*  fileURL = [panel URL];
+                    
+                    NSBitmapImageRep *bitmapRep = nil;
+                    for (NSImageRep *imageRep in [image representations]) {
+                        if ([imageRep isKindOfClass:[NSBitmapImageRep class]]){
+                            bitmapRep = (NSBitmapImageRep *)imageRep;
+                            break; // stop on first bitmap rep we find
+                        }
+                    }
+                    
+                    if (!bitmapRep) {
+                        bitmapRep = [NSBitmapImageRep imageRepWithData:[image TIFFRepresentation]];
+                        
+                    }
+                    
+                    NSString *extension = [[fileURL pathExtension] lowercaseString];
```

Comment:
File name extensions aren’t especially reliable on Apple platforms, since a file’s canonical type is represented by a [UTI](https://developer.apple.com/library/content/documentation/Miscellaneous/Reference/UTIRef/Articles/System-DeclaredUniformTypeIdentifiers.html). The save panel even lets you save without an extension unless you restrict the allowed file types to those that can be represented as images:

```objc
panel.allowedFileTypes = [NSImage imageTypes];
```

Here’s [a more robust way](https://stackoverflow.com/a/1506350/4585461) to go from a file extension to a UTI.

Reply:
An alternative to looking at file extensions:

1. The save panel tells us the file path. use NSFileManager to [determine the UTI at that path](https://stackoverflow.com/questions/25498417/how-do-i-get-a-document-uti-from-a-file-path#comment39847300_25498566). (Not sure if this only works for existing files or if it would work for unsaved files as well.)
1. Use `UTTypeConformsTo()` to determine whether the UTI at that path conforms to `public.jpeg` etc.
1. Avoid hard-coding UTIs like `public.jpeg` by iterating over `+[NSImage imageTypes]`.